integrations/jta/jdbc/src/main/java/io/helidon/integrations/jta/jdbc/LocalXAResource.java

```diff
@@ -0,0 +1,1031 @@
+        if (a != null) {
+            throw new UncheckedXAException((XAException) new XAException(XAER_DUPID)
+                                           .initCause(new IllegalArgumentException("xid: " + x + "; association: " + a)));
+        }
+        Connection c;
+        try {
+            c = this.connectionFunction.apply(x);
+        } catch (RuntimeException e) {
+            // Weirdly, XAER_RMERR seems to be the "please retry" error code in this one case, not XAER_RMFAIL:
+            // https://github.com/jbosstm/narayana/blob/8ccaf0f85c7a76c227941d26cc3aa3fa9f05b160/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/TransactionImple.java#L682-L689.
+            throw new UncheckedXAException((XAException) new XAException(XAER_RMERR).initCause(e));
+        }
+        if (c == null) {
+            // Weirdly, XAER_RMERR seems to be the "please retry" error code in this one case, not XAER_RMFAIL:
+            // https://github.com/jbosstm/narayana/blob/8ccaf0f85c7a76c227941d26cc3aa3fa9f05b160/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/TransactionImple.java#L682-L689.
+            throw new UncheckedXAException((XAException) new XAException(XAER_RMERR)
+                                           .initCause(new NullPointerException("connectionFunction.apply(" + x + ")")));
+        }
+        a = new Association(Association.BranchState.ACTIVE, x, c);
+        if (LOGGER.isLoggable(Level.FINE)) {
+            LOGGER.logp(Level.FINE, this.getClass().getName(), "start",
+                        "Created new Association ({0}) for connection ({1}) in state ACTIVE", new Object[] {a, c});
+        }
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.exiting(this.getClass().getName(), "start", a);
+        }
+        return a;
+    }
+
+    @Override // XAResource
+    public void end(Xid xid, int flags) throws XAException {
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.entering(this.getClass().getName(), "end", new Object[] {xid, flagsToString(flags)});
+        }
+        requireNonNullXid(xid);
+
+        BiFunction<Xid, Association, Association> remappingFunction;
+        switch (flags) {
+        case TMFAIL:
+        case TMSUCCESS:
+            remappingFunction = LocalXAResource::activeToIdle;
+            break;
+        case TMSUSPEND:
+            remappingFunction = LocalXAResource::suspend;
+            break;
+        default:
+            // Bad flags.
+            throw (XAException) new XAException(XAER_INVAL)
+                .initCause(new IllegalArgumentException("xid: " + xid + "; flags: " + flagsToString(flags)));
+        }
+
+        // Any XAException thrown can have any error code. The transaction will be marked as rollback only. See
+        // https://github.com/jbosstm/narayana/blob/8ccaf0f85c7a76c227941d26cc3aa3fa9f05b160/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/TransactionImple.java#L978-L992.
```

Comment:
Once again, figuring out what kind of error code to throw when is trial and error in the real world.